kernel/src/hil/gpio.rs

```diff
@@ -1,67 +1,326 @@
 
 /// Enum for configuring any pull-up or pull-down resistors on the GPIO pin.
 #[derive(Debug)]
-pub enum InputMode {
+pub enum FloatingState {
     PullUp,
     PullDown,
     PullNone,
 }
 
 /// Enum for selecting which edge to trigger interrupts on.
 #[derive(Debug)]
-pub enum InterruptMode {
+pub enum InterruptEdge {
     RisingEdge,
     FallingEdge,
     EitherEdge,
 }
 
-pub trait PinCtl {
-    /// Configure whether the pin should have a pull-up or pull-down resistor or
-    /// neither.
-    fn set_input_mode(&self, _: InputMode);
+/// Enum for which state the pin is in. Some MCUs can support Input/Output pins,
+/// so this is a valid option. `Function` means the pin has been configured to
+/// a special function. Determining which function it outside the scope of the HIL,
+/// and should instead use a chip-specific API.
+#[derive(Debug)]
+pub enum Configuration {
+    /// Cannot be read or written or used; effectively inactive.
+    LowPower,
+    /// Calls to the `Input` trait are valid.
+    Input,
+    /// Calls to the `Output` trait are valid.
+    Output,
+    /// Calls to both the `Input` and `Output` traits are valid.
+    InputOutput,
+    /// Chip-specific, requires chip-specific API for more detail,
+    Function,
+    /// In a state not covered by other values.
+    Other,
 }
 
-/// Interface for synchronous GPIO pins.
-pub trait Pin {
-    /// Configure the GPIO pin as an output pin.
-    fn make_output(&self);
+/// The Pin trait allows a pin to be used as either input
+/// or output and to be configured.
+pub trait Pin: Input + Output + Configure {}
+
+/// The InterruptPin trait allows a pin to be used as either
+/// input or output and also to source interrupts.
+pub trait InterruptPin: Pin + Interrupt {}
+
+/// The InterruptValuePin trait allows a pin to be used as
+/// either input or output and also to source interrupts which
+/// pass a value.
+pub trait InterruptValuePin: Pin + InterruptWithValue {}
+/// Control and configure a GPIO pin.
+pub trait Configure {
+    /// Return the current pin configuration.
+    fn configuration(&self) -> Configuration;
+
+    /// Make the pin an output, returning the current configuration,
+    /// which should be either `Configuration::Output` or
+    /// `Configuration::InputOutput`.
+    fn make_output(&self) -> Configuration;
+    /// Disable the pin as an output, returning the current configuration.
+    fn disable_output(&self) -> Configuration;
+
+    /// Make the pin an input, returning the current configuration,
+    /// which should be ither `Configuration::Input` or
+    /// `Configuration::InputOutput`.
+    fn make_input(&self) -> Configuration;
+    /// Disable the pin as an input, returning the current configuration.
+    fn disable_input(&self) -> Configuration;
+
+    /// Disable the pin and put it into its lowest power state.
+    /// Re-enabling the pin requires reconfiguring it (state of
+    /// its enabled configuration is not stored).
+    fn low_power(&self);
```

Comment:
What about making this (and I can push the commit) `deactivate_to_low_power()` (or something)? I think that seeing a call like `my_pin.low_power()` won't be intuitive that that actually stops the pin from working unless you have studied this HIL. Also, it seems like you would always want to call `my_pin.low_power()` because who would want a high power pin? And since the function doesn't guarantee to actually _disable_ the pin, "deactivate" seemed like an appropriate verb to signal that things likely won't work after calling this function.

Reply:
Yeah, that makes sense. Or `put_in_low_power()`.